aws-lc-rs/src/key_wrap.rs

```diff
@@ -0,0 +1,424 @@
+    /// The key-encryption key used with the selected cipher algorithn to wrap or unwrap a key.
+    pub struct KeyEncryptionKey<Cipher: BlockCipher> {
+        cipher: &'static Cipher,
+        key: Box<[u8]>,
+    }
+
+    impl<Cipher: BlockCipher> KeyEncryptionKey<Cipher> {
+        /// Construct a new Key Encryption Key.
+        ///
+        /// # Errors
+        /// * [`Unspecified`]: Any error that occurs constructing the key encryption key.
+        pub fn new(cipher: &'static Cipher, key: &[u8]) -> Result<Self, Unspecified> {
+            if key.len() != cipher.key_len() {
+                return Err(Unspecified);
+            }
+
+            let key = Vec::from(key).into_boxed_slice();
+
+            Ok(Self { cipher, key })
+        }
+
+        /// Returns the block cipher algorithm identifier configured for the key.
+        #[must_use]
+        pub fn block_cipher_id(&self) -> BlockCipherId {
+            self.cipher.id()
+        }
+    }
+
+    impl<Cipher: BlockCipher> Sealed for KeyEncryptionKey<Cipher> {}
+
+    impl KeyWrap for KeyEncryptionKey<AesBlockCipher> {
+        /// Peforms the key wrap encryption algorithm using `KeyEncryptionKey`'s configured block cipher.
+        /// It wraps `plaintext` and writes the corresponding ciphertext to `output`.
+        ///
+        /// # Validation
+        /// * `plaintext.len()` must be a multiple of eight
+        /// * `output.len() >= (input.len() + 8)`
+        ///
+        /// # Errors
+        /// * [`Unspecified`]: An error occurred either due to `output` being insufficiently sized, `input` exceeding
+        /// the allowed input size, or for other unspecified reasons.
+        fn wrap<'output>(
+            self,
+            plaintext: &[u8],
+            output: &'output mut [u8],
+        ) -> Result<&'output mut [u8], Unspecified> {
+            if output.len() < plaintext.len() + 8 {
+                return Err(Unspecified);
+            }
+
+            let mut aes_key = MaybeUninit::<AES_KEY>::uninit();
+
+            let key_bits: u32 = (self.key.len() * 8).try_into().map_err(|_| Unspecified)?;
+
+            if 0 != unsafe {
+                AES_set_encrypt_key(self.key.as_ptr(), key_bits, aes_key.as_mut_ptr())
+            } {
+                return Err(Unspecified);
+            }
+
+            let aes_key = unsafe { aes_key.assume_init() };
+
+            // AWS-LC validates the following:
+            // * in_len <= INT_MAX - 8
+            // * in_len >= 16
+            // * in_len % 8 == 0
+            let out_len = indicator_check!(unsafe {
+                AES_wrap_key(
+                    &aes_key,
+                    null(),
+                    output.as_mut_ptr(),
+                    plaintext.as_ptr(),
+                    plaintext.len(),
+                )
+            });
+
+            if out_len == -1 {
+                return Err(Unspecified);
+            }
+
+            let out_len: usize = out_len.try_into().map_err(|_| Unspecified)?;
```

Comment:
`AES_wrap_key` returns an int rather then a size_t, so on the Rust size we want to make sure we can fit an `i32` into a `usize` type for that architecture. So the `try_into` safe-guards this cast so that it fails if it would truncate.